### ISSUE TITLE
[Bug]: Regression in rrweb style capture for material ui

### ISSUE THREAD

**Issue by arijit91** [rrweb-io/rrweb#928]
### Preflight Checklist

- [X] I have searched the [issue tracker](https://www.github.com/rrweb-io/rrweb/issues) for a bug report that matches the one I want to file, without success.

### What package is this bug report for?

rrweb

### Expected Behavior

rrweb is not capturing styling changes coming from material-ui

Sample app (button color is expected to change from red to blue on click):
https://ckn2h.csb.app/#

This was previously reported and fixed in
https://github.com/rrweb-io/rrweb/issues/650

It is broken on main

I did a binary search and looks like this was the commit that broke it:
https://github.com/rrweb-io/rrweb/commit/a01c97f70da6583400d2759f26a518037d4c2adf

### Actual Behavior

Sample app (button color is expected to change from red to blue on click):
https://ckn2h.csb.app/#

**Comment by Yuyz0112** [comment]
Do you think https://github.com/rrweb-io/rrweb/commit/9ed6767adac4e0afab66fc3fb9716ab87bfc81f8 fixed this?

**Comment by arijit91** [comment]
@Yuyz0112 I thought so as well actually, I tried that patch but it didn't seem to work :(

The bug exists on main

**Comment by Yuyz0112** [comment]
ok, going to debug it

**Comment by arijit91** [comment]
From what I could make out, there is a css rule we pick up on with the css style declaration observer, we add it to the node, and then we try to append a node with the same text content from a DOM mutation, so we have 2 (duplicate) styles on the same node

Then while replaying the DOM mutation, only one of the css rules is updated..

**Comment by Yuyz0112** [comment]
you are right, the issue was introduced in https://github.com/rrweb-io/rrweb/commit/a01c97f70da6583400d2759f26a518037d4c2adf. And I'm sorry I did not add enough information about what it's trying to fix.

Now let us add comprehensive test cases for the stylesheet recording to ensure we handle things correctly without introducing further regression.

- remote link element, use href to load a stylesheet
- style element, apply stylesheet as text content inside it
  - existed when snapshot starting
  - dynamic created and recorded by observers
- style element, apply stylesheet by using CSS APIs, without text content inside it
  - existed when snapshot starting
  - dynamic created and recorded by observers

I think some duplication of case 2 and case 3 leads these issues, https://github.com/rrweb-io/rrweb/commit/9ed6767adac4e0afab66fc3fb9716ab87bfc81f8 fixed part of it, but not all.

**Comment by Yuyz0112** [comment]
@Juice10 are there any cases I'm missing here?

I think maybe we can only use CSS APIs' observers to record style, without serializing style elements' text content.

**Comment by Juice10** [comment]
@Yuyz0112 I have seen some combinations of the the last two cases you mentioned.

Also some style element can have multiple text nodes added inside them (via JS), which can lead to some unexpected behavior.

And another thing to keep in mind are some cssRules (like media queries) can have more cssRules inside them. It's possible that serializing isn't accounting for this and those rules aren't correctly captured.
